src/algebra/char_p/basic.lean

```diff
@@ -17,7 +14,39 @@ import ring_theory.nilpotent
 
 universes u v
 
-variables (R : Type u)
+open_locale big_operators
+
+variables {R : Type*}
+
+section semiring
+variables [semiring R] {p : ℕ} {x y : R}
+
+lemma commute.exists_add_pow_prime_pow_eq (hp : p.prime) (h : commute x y) (n : ℕ) :
+  ∃ r : R, (x + y) ^ p ^ n = x ^ p ^ n + y ^ p ^ n + p * r :=
+begin
+  refine ⟨∑ k in finset.range (p ^ n - 1),
+    x ^ (k + 1) * y ^ (p ^ n - (k + 1)) * ↑((p ^ n).choose (k + 1) / p), _⟩,
+  rw [h.add_pow, finset.sum_range_succ_comm, ←nat.sub_add_cancel (pow_pos hp.pos _),
+    finset.sum_range_succ'],
+  simp only [nat.sub_add_cancel (pow_pos hp.pos _), tsub_zero, mul_one, one_mul,
+    nat.choose_zero_right, nat.cast_one, pow_zero, tsub_self, nat.choose_self],
+  rw [add_comm _ (y ^ _), ←add_assoc, finset.mul_sum, finset.sum_congr rfl (λ i hi, _)],
+  rw [nat.cast_comm, mul_assoc, mul_assoc, mul_assoc, ←nat.cast_mul, nat.div_mul_cancel],
+  rw [finset.mem_range, lt_tsub_iff_right] at hi,
+  exact hp.dvd_choose_pow i.succ_ne_zero hi,
+end
+
+lemma commute.exists_add_pow_prime_eq (hp : p.prime) (h : commute x y) :
+  ∃ r : R, (x + y) ^ p = x ^ p + y ^ p + p * r :=
+by simpa using h.exists_add_pow_prime_pow_eq hp 1
+
+end semiring
+
+lemma add_pow_prime_eq_pow_add_pow_add_prime_mul [comm_semiring R] {p : ℕ} (hp : p.prime)
```

Comment:
Should this have an `exists` in the name? Or be stated without the exists at all, as above?